Sources/PackageModel/Toolset.swift

```diff
@@ -0,0 +1,214 @@
+import struct TSCBasic.AbsolutePath
+import protocol TSCBasic.FileSystem
+import struct TSCBasic.RelativePath
+import struct TSCBasic.StringError
+import struct TSCUtility.Version
+
+/// A set of paths and flags for tools used for building Swift packages. This type unifies pre-existing assorted ways
+/// to specify these properties across SwiftPM codebase.
+public struct Toolset {
+    public enum KnownTool: String, Hashable, CaseIterable {
+        case swiftCompiler
+        case cCompiler
+        case cxxCompiler
+        case linker
+        case librarian
+        case debugger
+    }
+
+    /// Properties of a known tool in a ``Toolset``.
+    public struct ToolProperties: Equatable {
+        /// Absolute path to the tool on the filesystem. If absent, implies a default tool is used.
+        public fileprivate(set) var path: AbsolutePath?
+
+        /// Command-line options to be passed to the tool when it's invoked.
+        public fileprivate(set) var extraCLIOptions: [String]?
+    }
+
+    /// A dictionary of known tools in this toolset.
+    public fileprivate(set) var knownTools: [KnownTool: ToolProperties]
+}
+
+extension Toolset {
+    /// Initialize a toolset from an encoded file on a file system.
+    /// - Parameters:
+    ///   - path: absolute path on the `fileSystem`.
+    ///   - fileSystem: file system from which the toolset should be read.
+    ///   - observability: an instance of `ObservabilityScope` to log warnings about unknown or invalid tools.
+    public init(from toolsetPath: AbsolutePath, at fileSystem: FileSystem, _ observability: ObservabilityScope) throws {
+        let decoder = JSONDecoder()
+        let decoded = try decoder.decode(path: toolsetPath, fileSystem: fileSystem, as: DecodedToolset.self)
+        guard decoded.schemaVersion == Version(1, 0, 0) else {
+            throw StringError(
+                "Unsupported `schemaVersion` \(decoded.schemaVersion) in toolset configuration at \(toolsetPath)"
+            )
+        }
+
+        var knownTools = [KnownTool: ToolProperties]()
+        for (tool, properties) in decoded.tools {
+            guard let knownTool = KnownTool(rawValue: tool) else {
+                observability.emit(warning: "Unknown tool `\(tool)` in toolset configuration at `\(toolsetPath)`")
+                continue
+            }
+
+            let toolPath: AbsolutePath?
+            if let path = properties.path {
+                if let absolutePath = try? AbsolutePath(validating: path) {
+                    toolPath = absolutePath
+                } else {
+                    let rootPath = decoded.rootPath ?? toolsetPath.parentDirectory
+                    toolPath = rootPath.appending(RelativePath(path))
+                }
+            } else {
+                toolPath = nil
+            }
+
+            guard toolPath != nil || !(properties.extraCLIOptions?.isEmpty ?? true) else {
+                // don't keep track of a tool with no path and CLI options specified.
```

Comment:
I think this should just be an error tbh

Reply:
That's updated now, it emits errors on the observability scope for every empty tool configuration and throws an error after that.